File: engine-rs/src/utils/entity_codex.rs
//! R10 实体卡 Codex 化（375 号契约批，三轮 P0）。
//!
//! TS 真源：`packages/core/src/utils/entity-codex.ts`；golden 唯一事实源：
//! `packages/core/src/__tests__/golden/entity-codex-vectors.json`（差分测试
//! `tests/golden_entity_codex_diff.rs` 读同一文件）。
//!
//! 实体卡 = 名册条目的视图扩展（同源派生，非第二真相源）：身份字段
//! （name/aliases/kind）以名册为准，卡片仅承载 summary/facts/relationships
//! 作者扩展。命中计数/排序与 TS 逐字一致（禁 locale 感知排序）。

use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexRosterEntity {
    #[serde(default)]
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub aliases: Vec<String>,
    #[serde(default = "default_kind")]
    pub kind: String,
    #[serde(default)]
    pub registered_at: i64,
}

fn default_kind() -> String {
    "other".to_string()
}

#[derive(Debug, Clone, PartialEq, serde::Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexRelationship {
    pub target: String,
    pub note: String,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EntityCodexCard {
    pub name: String,
    #[serde(default)]
    pub aliases: Vec<String>,
    #[serde(default)]
    pub kind: String,
    #[serde(default)]
    pub summary: String,
    #[serde(default)]
    pub facts: Vec<String>,
    #[serde(default)]
    pub relationships: Vec<CodexRelationship>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub first_chapter: Option<i64>,
}

/// 名册 → 卡片视图（同源派生）：summary/facts 置空，按 name 码元序排序。
pub fn derive_codex_cards(roster: &[CodexRosterEntity]) -> Vec<EntityCodexCard> {
    let mut cards: Vec<EntityCodexCard> = roster
        .iter()
        .map(|entity| EntityCodexCard {
            name: entity.name.clone(),
            aliases: entity.aliases.clone(),
            kind: entity.kind.clone(),
            summary: String::new(),
            facts: Vec::new(),
            relationships: Vec::new(),
            first_chapter: None,
        })
        .collect();
    cards.sort_by(|a, b| a.name.cmp(&b.name));
    cards
}

const CODEX_LIMITS_SUMMARY: usize = 300;
const CODEX_LIMITS_FACTS: usize = 10;
const CODEX_LIMITS_FACT: usize = 200;
const CODEX_LIMITS_RELATIONSHIPS: usize = 10;
const CODEX_LIMITS_RELATIONSHIP_NOTE: usize = 200;

fn clamp_list(items: &[String], max_items: usize, max_chars: usize) -> Vec<String> {
    items
        .iter()
        .map(|item| item.trim().to_string())
        .filter(|item| !item.is_empty())
        .take(max_items)
        .map(|item| {
            if item.chars().count() > max_chars {
                item.chars().take(max_chars).collect()
            } else {
                item
            }
        })
        .collect()
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexRevision {
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub facts: Option<Vec<String>>,
    #[serde(default)]
    pub relationships: Option<Vec<CodexRelationship>>,
}

/// 卡片修订：以名册同源卡为底座合并作者扩展（summary/facts/relationships）。
pub fn revise_codex_card(base: &EntityCodexCard, revision: &CodexRevision) -> EntityCodexCard {
    let summary = revision
        .summary
        .clone()
        .unwrap_or_else(|| base.summary.clone());
    let summary = if summary.chars().count() > CODEX_LIMITS_SUMMARY {
        summary.chars().take(CODEX_LIMITS_SUMMARY).collect()
    } else {
        summary
    };
    EntityCodexCard {
        name: base.name.clone(),
        aliases: base.aliases.clone(),
        kind: base.kind.clone(),
        summary,
        facts: clamp_list(
            revision.facts.as_ref().unwrap_or(&base.facts),
            CODEX_LIMITS_FACTS,
            CODEX_LIMITS_FACT,
        ),
        relationships: {
            let relationships = revision
                .relationships
                .as_ref()
                .unwrap_or(&base.relationships);
            relationships
                .iter()
                .take(CODEX_LIMITS_RELATIONSHIPS)
                .map(|relation| CodexRelationship {
                    target: relation.target.clone(),
                    note: {
                        let note = if relation.note.chars().count() > CODEX_LIMITS_RELATIONSHIP_NOTE {
                            relation.note.chars().take(CODEX_LIMITS_RELATIONSHIP_NOTE).collect()
                        } else {
                            relation.note.clone()
                        };
                        note
                    },
                })
                .collect()
        },
        first_chapter: base.first_chapter,
    }
}

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexMatch {
    pub card: EntityCodexCard,
    /// 命中次数（name 与 alias 合并计）。
    pub hits: i64,
}

/// 场景命中：text 含 name 或任一 alias 的卡片（0 命中不返回）。
pub fn match_codex_cards(text: &str, cards: &[EntityCodexCard]) -> Vec<CodexMatch> {
    let mut matches: Vec<CodexMatch> = Vec::new();
    for card in cards {
        let mut hits: i64 = 0;
        for term in std::iter::once(&card.name).chain(card.aliases.iter()) {
            if term.is_empty() {
                continue;
            }
            let term_chars = term.chars().count().max(1);
            let mut from = 0usize;
            let char_len = text.chars().count();
            while from <= char_len {
                let hay: String = text.chars().skip(from).collect();
                match hay.find(term.as_str()) {
                    Some(relative) => {
                        hits += 1;
                        from += hay[..relative].chars().count() + term_chars;
                    }
                    None => break,
                }
            }
        }
        if hits > 0 {
            matches.push(CodexMatch { card: card.clone(), hits });
        }
    }
    matches.sort_by(|a, b| {
        b.hits
            .cmp(&a.hits)
            .then_with(|| a.card.name.cmp(&b.card.name))
    });
    matches
}

const CODEX_EXCERPT_MAX: usize = 400;

/// 注入渲染：命中卡片 → 「## 实体卡」块（无命中返回 None）。
pub fn render_codex_block(
    matches: &[CodexMatch],
    language: crate::utils::language::WritingLanguage,
) -> Option<String> {
    if matches.is_empty() {
        return None;
    }
    let is_en = language == crate::utils::language::WritingLanguage::En;
    let sections: Vec<String> = matches
        .iter()
        .map(|codex_match| {
            let card = &codex_match.card;
            let mut lines = Vec::new();
            if card.aliases.is_empty() {
                lines.push(format!("### {}", card.name));
            } else {
                lines.push(format!("### {}（{}）", card.name, card.aliases.join("、")));
            }
            if !card.summary.is_empty() {
                if card.summary.chars().count() > CODEX_EXCERPT_MAX {
                    let clipped: String =
                        card.summary.chars().take(CODEX_EXCERPT_MAX).collect();
                    lines.push(format!("{clipped}…"));
                } else {
                    lines.push(card.summary.clone());
                }
            }
            if !card.facts.is_empty() {
                lines.push(if is_en { "Canon facts:".to_string() } else { "正典事实：".to_string() });
                for fact in &card.facts {
                    lines.push(format!("- {fact}"));
                }
            }
            if !card.relationships.is_empty() {
                lines.push(if is_en { "Relationships:".to_string() } else { "关系：".to_string() });
                for relation in &card.relationships {
                    lines.push(format!("- {}: {}", relation.target, relation.note));
                }
            }
            lines.join("\n")
        })
        .collect();
    let header = if is_en {
        "## Entity cards (scene cast — follow canon facts)"
    } else {
        "## 实体卡（本场出场——遵循正典事实）"
    };
    let mut parts = vec![header.to_string()];
    parts.extend(sections);
    Some(parts.join("\n"))
}

File: packages/core/src/utils/entity-codex.ts
/**
 * R10 实体卡 Codex 化（375 号契约批，三轮 P0；v3 §3 R10）。
 *
 * 实体卡 = 名册条目的**视图扩展**（同源派生，非第二真相源）：名册
 * （G7b entity_roster.md）仍是唯一身份真相，卡片在视图层补充摘要/
 * 硬事实/关系供写作注入。核心契约：
 *
 * 1. `deriveCodexCards`：名册 → 卡片视图（同源派生，作者编辑只发生在
 *    卡片扩展字段上，name/aliases/kind 永远以名册为准）；
 * 2. `matchCodexCards`：场景文本（正文/备忘）按 name+aliases 命中卡片，
 *    命中数降序稳定排序（禁 localeCompare）；
 * 3. `renderCodexBlock`：命中卡片 → 「## 实体卡」注入块（双语）。
 *
 * 双端：`engine-rs/src/utils/entity_codex.rs` 1:1 镜像；共享向量
 * `src/__tests__/golden/entity-codex-vectors.json`（TS 断言
 * `golden-entity-codex.test.ts`；Rust 差分 `tests/golden_entity_codex_diff.rs`）。
 */

import type { RosterEntity } from "./entity-roster.js";

export interface CodexRelationship {
  readonly target: string;
  readonly note: string;
}

export interface EntityCodexCard {
  /** 与名册 name 一致（同源键）。 */
  readonly name: string;
  readonly aliases: ReadonlyArray<string>;
  readonly kind: RosterEntity["kind"];
  /** 卡片摘要（谁/什么，≤300 码元；名册派生时为空，作者可编辑）。 */
  readonly summary: string;
  /** 硬事实（正典约束，≤10 条 × ≤200 码元）。 */
  readonly facts: ReadonlyArray<string>;
  /** 关系（≤10 条）。 */
  readonly relationships: ReadonlyArray<CodexRelationship>;
  /** 首次出场章（名册 registeredAt 不可用时缺省）。 */
  readonly firstChapter?: number;
}

export const CODEX_LIMITS = {
  summary: 300,
  facts: 10,
  fact: 200,
  relationships: 10,
  relationshipNote: 200,
} as const;

/** 名册 → 卡片视图（同源派生）：summary/facts 置空，身份字段以名册为准。 */
export function deriveCodexCards(roster: ReadonlyArray<RosterEntity>): EntityCodexCard[] {
  return roster
    .map((entity) => ({
      name: entity.name,
      aliases: [...entity.aliases],
      kind: entity.kind,
      summary: "",
      facts: [] as string[],
      relationships: [] as CodexRelationship[],
    }))
    .sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
}

const clampList = (items: ReadonlyArray<string>, maxItems: number, maxChars: number): string[] =>
  items
    .map((item) => item.trim())
    .filter(Boolean)
    .slice(0, maxItems)
    .map((item) => (item.length > maxChars ? item.slice(0, maxChars) : item));

/**
 * 518 号：**消费面**卡归一化（Rust 侧 EntityCodexCard 全字段 `#[serde(default)]`
 * 的 TS 对应物；读写端点双端均 Value 原样透传，容差只发生在进管线前）。
 * 缺 aliases 的卡此前在 matchCodexCards 展开 `...card.aliases` 即崩
 * （差分器实证：PUT codex 200 → resync 500）。未知字段（如 Rust 回显的 id）
 * spread 保留；kind 仅补缺省不校验取值（Rust 侧为 String 无枚举约束）。
 */
export function normalizeCodexCards(cards: ReadonlyArray<unknown>): EntityCodexCard[] {
  const out: EntityCodexCard[] = [];
  for (const card of cards) {
    if (!card || typeof card !== "object") continue;
    const raw = card as Record<string, unknown>;
    if (typeof raw.name !== "string" || raw.name.length === 0) continue;
    out.push({
      ...raw,
      name: raw.name,
      aliases: Array.isArray(raw.aliases)
        ? raw.aliases.filter((alias): alias is string => typeof alias === "string")
        : [],
      kind: (typeof raw.kind === "string" ? raw.kind : "other") as EntityCodexCard["kind"],
      summary: typeof raw.summary === "string" ? raw.summary : "",
      facts: Array.isArray(raw.facts)
        ? raw.facts.filter((fact): fact is string => typeof fact === "string")
        : [],
      relationships: Array.isArray(raw.relationships)
        ? raw.relationships.filter(
            (relation): relation is CodexRelationship =>
              !!relation && typeof relation === "object"
                && typeof (relation as CodexRelationship).target === "string"
                && typeof (relation as CodexRelationship).note === "string",
          )
        : [],
    } as EntityCodexCard);
  }
  return out;
}

/** 卡片修订：以名册同源卡为底座合并作者扩展（summary/facts/relationships）。 */
export function reviseCodexCard(
  base: EntityCodexCard,
  revision: Partial<Pick<EntityCodexCard, "summary" | "facts" | "relationships">>,
): EntityCodexCard {
  return {
    ...base,
    summary: (revision.summary ?? base.summary).slice(0, CODEX_LIMITS.summary),
    facts: clampList(revision.facts ?? base.facts, CODEX_LIMITS.facts, CODEX_LIMITS.fact),
    relationships: (revision.relationships ?? base.relationships)
      .slice(0, CODEX_LIMITS.relationships)
      .map((relation) => ({
        target: relation.target,
        note: relation.note.slice(0, CODEX_LIMITS.relationshipNote),
      })),
  };
}

export interface CodexMatch<T> {
  readonly card: T;
  /** 命中次数（name 与 alias 合并计）。 */
  readonly hits: number;
}

/** 场景命中：text 含 name 或任一 alias 的卡片（0 命中不返回）。 */
export function matchCodexCards<T extends EntityCodexCard>(
  text: string,
  cards: ReadonlyArray<T>,
): Array<CodexMatch<T>> {
  const matches: Array<CodexMatch<T>> = [];
  for (const card of cards) {
    let hits = 0;
    for (const term of [card.name, ...card.aliases]) {
      if (!term) continue;
      let from = 0;
      while (from <= text.length) {
        const index = text.indexOf(term, from);
        if (index < 0) break;
        hits += 1;
        from = index + Math.max(1, term.length);
      }
    }
    if (hits > 0) matches.push({ card, hits });
  }
  return matches.sort(
    (a, b) =>
      b.hits - a.hits ||
      (a.card.name < b.card.name ? -1 : a.card.name > b.card.name ? 1 : 0),
  );
}

const CODEX_EXCERPT_MAX = 400;

/** 注入渲染：命中卡片 → 「## 实体卡」块（无命中返回 undefined）。 */
export function renderCodexBlock(
  matches: ReadonlyArray<CodexMatch<EntityCodexCard>>,
  language: "zh" | "en" = "zh",
): string | undefined {
  if (matches.length === 0) return undefined;
  const isEn = language === "en";
  const sections = matches.map(({ card }) => {
    const aliasPart = card.aliases.length > 0 ? `（${card.aliases.join("、")}）` : "";
    const lines = [`### ${card.name}${aliasPart}`];
    if (card.summary) {
      lines.push(card.summary.length > CODEX_EXCERPT_MAX ? `${card.summary.slice(0, CODEX_EXCERPT_MAX)}…` : card.summary);
    }
    if (card.facts.length > 0) {
      lines.push(isEn ? "Canon facts:" : "正典事实：");
      for (const fact of card.facts) lines.push(`- ${fact}`);
    }
    if (card.relationships.length > 0) {
      lines.push(isEn ? "Relationships:" : "关系：");
      for (const relation of card.relationships) {
        lines.push(`- ${relation.target}: ${relation.note}`);
      }
    }
    return lines.join("\n");
  });
  return [
    isEn ? "## Entity cards (scene cast — follow canon facts)" : "## 实体卡（本场出场——遵循正典事实）",
    ...sections,
  ].join("\n");
}
